File: DDDownload/DDDownload/DDDownload/DDDownloadManager.m
//
//  DDDownloadManager.m
//  DDDownload
//
//  Created by wuqh on 2018/12/13.
//  Copyright © 2018 wuqh. All rights reserved.
//

#import "DDDownloadManager.h"
#import "DDDownloadCache.h"
#import "NSString+md5.h"
#import "DDDownloadModel.h"

@interface DDDownloadManager()<NSURLSessionDownloadDelegate,NSURLSessionTaskDelegate>

@property(nonatomic, strong) NSURLSession *session;
@property(nonatomic, strong) NSMutableDictionary *downloadTasks;

// 下载文件夹，下载的视频都放在这里
@property(nonatomic, readonly) NSString *downloadDirectory;

@end

@implementation DDDownloadManager

static id _instance;
+ (instancetype)sharedManager {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _instance = [[DDDownloadManager alloc]init];
    });
    return _instance;
}

- (instancetype)init {
    if (self = [super init]) {
        [DDDownloadManager creatDirectorys];
        self.session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration backgroundSessionConfigurationWithIdentifier:@"com.DDDownloadManager"] delegate:self delegateQueue:[NSOperationQueue mainQueue]];
        self.downloadTasks = @{}.mutableCopy;
    }
    return self;
}

#pragma mark public method
- (void)downloadWithUrl:(NSString *)url {
    //1.查询是否加入了下载队列
    if (self.downloadTasks[url]) {
        return;
    }
    
    NSURLSessionDownloadTask *task;
    if ([DDDownloadCache.shared resumeDataForUrl:url]) {
        task = [self.session downloadTaskWithResumeData:[DDDownloadCache.shared resumeDataForUrl:url]];
    }else {
        task = [self.session downloadTaskWithURL:[NSURL URLWithString:url]];
    }
    self.downloadTasks[url] = task;
    
    if ([DDDownloadCache.shared downloadModelForUrl:url] == nil) {
        DDDownloadModel *downloadModel = [[DDDownloadModel alloc] init];
        [downloadModel setValue:url forKey:@"url"];
        [DDDownloadCache.shared setDownloadModel:downloadModel forUrl:url];
    }
    [task resume];
    
}
- (void)cancelDownloadWithUrl:(NSString *)url {
    NSURLSessionDownloadTask *task;
    if ((task = self.downloadTasks[url])) {
        [task cancelByProducingResumeData:^(NSData * _Nullable resumeData) {
            [DDDownloadCache.shared setResumeData:resumeData forUrl:url];
        }];
    }else {
        NSLog(@"取消异常，不可能走到这里");
    }
}

- (void)deleteDownloadWithUrl:(NSString *)url {
    //如果正在队列中请求，取消请求并删除
    NSURLSessionDownloadTask *task;
    if ((task = self.downloadTasks[url])) {
        [task cancel];
        [self.downloadTasks removeObjectForKey:url];
    }
    //删除resumeData
    [DDDownloadCache.shared removeResumeDataForUrl:url];
    //删除downloadModel
    [DDDownloadCache.shared removeDownloadModelForUrl:url];
    //如果文件下载成功，删除下载的文件
    if ([NSFileManager.defaultManager fileExistsAtPath:[DDDownloadManager downloadFilePathWithUrl:url]]) {
        [NSFileManager.defaultManager removeItemAtPath:[DDDownloadManager downloadFilePathWithUrl:url] error:nil];
    }
    
}

#pragma mark NSURLSessionDelegate
- (void)URLSession:(NSURLSession *)session
      downloadTask:(NSURLSessionDownloadTask *)downloadTask
didFinishDownloadingToURL:(NSURL *)location {
    
    NSString *url = downloadTask.currentRequest.URL.absoluteString;
    NSError *error;
    
    [NSFileManager.defaultManager moveItemAtURL:location toURL:[NSURL fileURLWithPath:[DDDownloadManager downloadFilePathWithUrl:url]] error:&error];
    if (error) {
        NSLog(@"%@",error);
        NSLog(@"完成异常，不可能走到这里");
    }

    
}
- (void)URLSession:(NSURLSession *)session
      downloadTask:(NSURLSessionDownloadTask *)downloadTask
      didWriteData:(int64_t)bytesWritten
 totalBytesWritten:(int64_t)totalBytesWritten
totalBytesExpectedToWrite:(int64_t)totalBytesExpectedToWrite {
    NSString *url = downloadTask.currentRequest.URL.absoluteString;
    
    DDDownloadModel *model = [DDDownloadCache.shared downloadModelForUrl:url];
    [model setValue:@(totalBytesWritten*1.0/totalBytesExpectedToWrite) forKey:@"progress"];
    [model setValue:@(totalBytesWritten) forKey:@"bytesWritten"];
    [model setValue:@(totalBytesExpectedToWrite) forKey:@"totalBytes"];
    
    [[NSNotificationCenter defaultCenter] postNotificationName:downloadTask.currentRequest.URL.absoluteString.md5 object:nil userInfo:@{@"downloadModel":model}];
}

- (void)URLSession:(NSURLSession *)session
      downloadTask:(NSURLSessionDownloadTask *)downloadTask
 didResumeAtOffset:(int64_t)fileOffset
expectedTotalBytes:(int64_t)expectedTotalBytes {
    
}

- (void)URLSession:(NSURLSession *)session
              task:(NSURLSessionTask *)task
didCompleteWithError:(nullable NSError *)error {
    
}

#pragma mark - 文件夹
/**
 创建存放下载文件的文件夹
 */
+ (void)creatDirectorys {
    [NSFileManager.defaultManager createDirectoryAtPath:DDDownloadManager.downloadDirectory withIntermediateDirectories:NO attributes:nil error:nil];
}

/**
 下载的文件夹地址

 @return 地址
 */
+ (NSString *)downloadDirectory {
    return [NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES).lastObject stringByAppendingString:@"/DDDownload/"];
}

/**
 根据url生成下载的文件存放地址

 @param url url
 @return 地址
 */
+ (NSString *)downloadFilePathWithUrl:(NSString *)url {
    return [DDDownloadManager.downloadDirectory stringByAppendingPathComponent:[NSString stringWithFormat:@"%@.%@",url.md5,url.pathExtension]];
}


@end

File: DDDownload/DDDownload/DDDownload/DDDownloadManager.swift
//
//  DDDownloadManager.swift
//  DDDownload
//
//  Created by wuqh on 2018/12/11.
//  Copyright © 2018 wuqh. All rights reserved.
//

import UIKit

class DDDownloadManager: NSObject {
    
    static let `default` = DDDownloadManager()
    
    var session: URLSession!
    var dataTasks = [String:URLSessionDownloadTask]()
    
    override init() {
        
        super.init()
        
        session = URLSession(configuration: URLSessionConfiguration.background(withIdentifier: "com.AIDownload"), delegate: self, delegateQueue: OperationQueue.main)
        
    }
    
}

extension DDDownloadManager {
    func setup() {
        
    }
}

extension DDDownloadManager {
    func download(withUrl url: String) {
        guard let URL = URL(string: url) else{
            return
        }
        
        if let data = DDDownloadManager.getResumeData(withUrl: url) {
            let dataTask = session.downloadTask(withResumeData: data)
            dataTasks[url] = dataTask
            dataTask.resume()
        }else {
            let dataTask = session.downloadTask(with: URL)
            dataTasks[url] = dataTask
            dataTask.resume()
        }
        
    }
    func cancelDownload(withUrl url: String) {
        dataTasks[url]?.cancel(byProducingResumeData: { (data) in
            if let resumeData = data {
                DDDownloadManager.setResuemData(withUrl: url, data: resumeData)
            }else {
                print("没有数据")
            }
        })
    }
}

extension DDDownloadManager: URLSessionDownloadDelegate {

    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didFinishDownloadingTo location: URL) {
        
        if let url = downloadTask.currentRequest?.url?.absoluteString {
            try? FileManager.default.moveItem(at: location, to: DDDownloadManager.downloadFilePath(url))
            try? FileManager.default.removeItem(atPath: DDDownloadManager.resumeDataFilePath(url).path)
        }

    }
    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didWriteData bytesWritten: Int64, totalBytesWritten: Int64, totalBytesExpectedToWrite: Int64) {
        
        if let url = downloadTask.currentRequest?.url?.absoluteString {
            
            if self.dataTasks[url] == nil {
                self.dataTasks[url] = downloadTask
                cancelDownload(withUrl: url)
            }
            
            NotificationCenter.default.post(name: NSNotification.Name.init(url.md5), object: nil, userInfo: ["url":url,"progress":Float(totalBytesWritten) / Float(totalBytesExpectedToWrite)])
            
            
        }
        
    }
//    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didResumeAtOffset fileOffset: Int64, expectedTotalBytes: Int64) {
//        print(#function)
//    }
    
    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        print(#function)
        
    }
}


// MARK: - resumeData操作相关
extension DDDownloadManager {
    static private func getResumeData(withUrl url: String) -> Data? {
        if FileManager.default.fileExists(atPath: resumeDataFilePath(url).path) {
            return try? Data(contentsOf: resumeDataFilePath(url))
        }else {
            return nil
        }
    }
    static private func setResuemData(withUrl url: String,data: Data) {
        try! data.write(to: resumeDataFilePath(url))
    }
}

// MARK: - 文件操作相关
extension DDDownloadManager {
    
    class var downloadDirectory: String {
        
        return String.init(NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).last!) + "/DDDownload/"
    }
    
    class var resumeDataDirectory: String {
        return String.init(NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).last!) + "/resumeData/"
    }
    
    static func downloadFilePath(_ fileName: String) -> URL {
        
        if FileManager.default.fileExists(atPath: DDDownloadManager.downloadDirectory) == false {
            try? FileManager.default.createDirectory(atPath: DDDownloadManager.downloadDirectory, withIntermediateDirectories: false, attributes: nil)
        }
        
        return URL(fileURLWithPath: DDDownloadManager.downloadDirectory + "\(fileName.md5)" + ".mp4")
    }
    static private func resumeDataFilePath(_ fileName: String) -> URL {
        if FileManager.default.fileExists(atPath: DDDownloadManager.resumeDataDirectory) == false {
            try? FileManager.default.createDirectory(atPath: DDDownloadManager.resumeDataDirectory, withIntermediateDirectories: false, attributes: nil)
        }
        return URL(fileURLWithPath: DDDownloadManager.resumeDataDirectory + "\(fileName.md5)")
    }
}
